File: C++/ArrayString/29_3SumWithMultiplicity.cpp
#include <bits/stdc++.h>
using namespace std;

int threeSumMulti(vector<int> &arr, int target)
{
    // sort(nums.begin(), nums.end());
    map<int, int> mp;
    sort(arr.begin(), arr.end());
    for (auto &i : arr)
        mp[i]++;
    int n = arr.size();
    vector<vector<int>> res;
    for (int k = 0; k < n; k++)
    {
        if (k > 0 && arr[k] == arr[k - 1])
            continue; // avoid duplicate as earlier we would have calculated every possibility of nums[k];
        int i = k + 1, j = n - 1;
        while (i < j)
        {
            int sum = arr[i] + arr[j] + arr[k];
            if (sum == target)
            {
                res.push_back({arr[k], arr[i], arr[j]});
                while (i < j && arr[i] == arr[i + 1])
                    i++;
                while (i < j && arr[j] == arr[j - 1])
                    j--;
                i++;
                j--;
            }
            else if (sum > target)
                j--;
            else
                i++;
        }
    }
    long ans = 0;
    for (auto &v : res)
    {
        if (v[0] == v[1] && v[1] == v[2])
        {
            //nc3
            long x = mp[v[0]];
            ans += x * (x - 1) * (x - 2) / 6;
        }
        else if (v[0] == v[1])
        {
            long x = mp[v[0]];
            ans += (x * (x - 1) / 2) * mp[v[2]];
        }
        else if (v[1] == v[2])
        {
            long x = mp[v[1]];
            ans += (x * (x - 1) / 2) * mp[v[0]];
        }
        else
            ans += mp[v[0]] * mp[v[1]] * mp[v[2]];
    }
    return ans % 1000000007;
}

File: Java/ArrayString/29_3SumWithMultiplicity.java
int mod = (int)1e9 + 7;
    public int twoSumMulti(int[] arr,int si,int ei, int tar) {
        long ans = 0;
        int n = arr.length;
        while(si < ei){
            int sum = arr[si] + arr[ei];
            if(sum < tar) si++;
            else if(sum > tar) ei--;
            else if(arr[si] != arr[ei]){
                int lc = 1, rc = 1;
                si++;ei--;
                while(si <= ei && arr[si] == arr[si-1]){
                    lc++;
                    si++;
                }
                while(si <= ei && arr[ei] == arr[ei + 1]){
                    rc++;
                    ei--;
                }
                ans += lc * rc;
                ans %= mod;
            }
            else{
                int N = ei - si + 1;
                ans += (N * (N-1))/2;
                ans %= mod;
                break;
            }
        }
        return (int)ans;
    }
    public int threeSumMulti(int[] arr, int tar) {
        int n = arr.length;
        Arrays.sort(arr);
        int ans =  0;
        for(int i=0;i<n-2;i++){
            ans = (ans + twoSumMulti(arr,i+1,n-1, tar - arr[i]));
            ans %= mod;
        }
        return ans;
    }